db/user_test.go

```diff
@@ -14,6 +15,88 @@ type SyncNewUsersTestCase struct {
     sendEmails bool
 }
 
+func (this *DBTests) DBTestResolveUsers(test *testing.T) {
+    defer ResetForTesting();
+
+    testCases := []struct {input []string; expectedOutput []string; addUsers []*model.User; removeUsers []string} {
+        // This is a simple test case for the empty string input.
+        {
+            []string{""},
+            []string{},
+            nil,
+            []string{},
+        },
+
+        // This is a test to ensure the output is sorted.
+        {
+            []string{"b@test.com", "a@test.com", "c@test.com"},
+            []string{"a@test.com", "b@test.com", "c@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a basic test to ensure that a role gets mapped to the correct email.
+        {
+            []string{"admin"},
+            []string{"admin@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a test for our all roles character, the *. 
+        {
+            []string{"*"},
+            []string{"admin@test.com", "grader@test.com", "other@test.com", "owner@test.com", "student@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a more complex test that adds multiple Users for the student role.
+        // This test case removes the only user from the "other" role, so we check that a role without any users still functions properly.
+        // This test case is given duplicate emails of different capitalizations and duplicate roles.
+        // It tests to ensures we do not produce duplicates on this input.
+        {
+            []string{"other", "*", "grader@test.com", "zoinks@test.com", "ZoinKS@teSt.Com"},
+            []string{"admin@test.com", "grader@test.com", "other@test.com", "second_student@test.com", "student@test.com", "zoinks@test.com"},
+            []*model.User{model.NewUser("second_student@test.com", "", model.GetRole("student"))},
+            []string{"owner@test.com"},
+        },
+
+        // This test case tests if miscapitalized roles still function and that we warn on invalid roles.
```

Comment:
There is no check for the warning, but there can be.
Just clear the logs and fetch warning logs.

You don't even need to check the contents, just the count should be sufficient.
Your test case struct can just have an int to represent the number of warnings created during the test case.